### PR TITLE
Document card serialization/deserialization

### PR DESCRIPTION
**Pull request by FadhlanR**

### CODE REVIEW

**Comment by habdelra** [comment]
please cover:
- custom serializers/deserializers and how those work (e.g. the datetime card in base is a good example)
- talk about relative URLs and how those are resolved when deserializing
- talk about how the `included` field works in the serialized card and how its handled as part of deserialization (which is probably a discussion around linksTo fields)
- talk about polymorphic fields and how those are serialized and deserialized
- talk about the 3 value types for cards: null, undefined, and actual value, and what that means in context of serialization and deserialization (aka the difference between empty and no value)
- talk about the plural fields and how we can serialize those with index numbers as part of the property path (https://github.com/cardstack/boxel/blob/4ec8af801231f9120911a13e31fb68f63ef264c7/packages/host/tests/integration/components/serialization-test.gts#L1432)
- talk about object identity and deserialization

File: docs/card-serialization-deserialization.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Card Serialization and Deserialization
+
+Card Serialization and Deserialization are essential processes that ensure a seamless transformation of card data between JSON format and actual card instances. We persistently store card data on the Realm server in two files: the card data is stored in a `.json` file, while the definition or class of the card is stored in a `.gts` file. Within the `.json` file, there is an `adoptFrom` field, where the value represents the path to the `.gts` file. These files are then accessed by the browser through the Realm server and served to the user.
```
Comment (lukemelia) [inline docs/card-serialization-deserialization.md]:
```suggestion
Card Serialization and Deserialization are essential processes that ensure a seamless transformation of card data between JSON format and actual card instances. We persistently store card data on the Realm server in two files: the card data is stored in a `.json` file, while the definition or class of the card is stored in a `.gts` file. Within the `.json` file, there is an `adoptsFrom` field, where the value represents the path to the `.gts` file. These files are then accessed by the browser through the Realm server and served to the user.
```

File: docs/card-serialization-deserialization.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+In the serialization process, we transform a card instance into a JSON object. This is achieved by gathering all the values of the card instance's fields from the data bucket.
+
+The central function employed in this process is `serializeCard`. This function takes a card instance as input and generates a `LooseSingleCardDocument` object, which can be directly converted into a JSON representation.
+
+## Card Deserialization
+
+In the deserialization process, we convert JSON data into a card instance. This enables the browser to utilize the instance for rendering a component to be presented to the user.
+
+We employ the `createFromSerialized` function for this purpose. This function triggers the creation of a card instance by retrieving the stored `.gts` file, which is accessible through the `adoptsFrom` field of the `LooseCardResource`. Subsequently, we iterate through the fields of the card instance and retrieve the value of each field from the `LooseCardResource`, storing it in the data bucket. During this process, we also execute the `computateVia` function to obtain the value for any computed fields.
```
Comment (lukemelia) [inline docs/card-serialization-deserialization.md]:
```suggestion
We employ the `createFromSerialized` function for this purpose. This function triggers the creation of a card instance by retrieving the stored `.gts` file, which is accessible through the `adoptsFrom` field of the `LooseCardResource`. Subsequently, we iterate through the fields of the card instance and retrieve the value of each field from the `LooseCardResource`, storing it in the data bucket. During this process, we also execute the `computeVia` function to obtain the value for any computed fields.
```

**Review by lukemelia — APPROVED** [review]
Good to go with my fixes. 👍

**Comment by tintinthong** [comment]
```mermaid
flowchart TB

    A[Card Instance] --> |serializeCard| B[LooseSingleCardDocument]
    B --> C
    E[LooseCardResource] -->|createFromSerialized| A
    C[JSON Storage on Realm Server] --> E

    style A fill:#58D68D
    style B fill:#3498DB
    style C fill:#F5B041
    style E fill:#3498DB
```

Does something like this make sense to help user visualise? I tried to tinker with mermaid but its kinda hard to enforce the position of nodes

**Review by tintinthong — APPROVED** [review]
Looks good

**Comment by FadhlanR** [comment]
> ```mermaid
> flowchart TB
> 
>     A[Card Instance] --> |serializeCard| B[LooseSingleCardDocument]
>     B --> C
>     E[LooseCardResource] -->|createFromSerialized| A
>     C[JSON Storage on Realm Server] --> E
> 
>     style A fill:#58D68D
>     style B fill:#3498DB
>     style C fill:#F5B041
>     style E fill:#3498DB
> ```
> Does something like this make sense to help user visualise? I tried to tinker with mermaid but its kinda hard to enforce the position of nodes

Nice diagram, I'll include it. Thanks!

**Pull request merged**